ephios/plugins/questionnaires/signals.py

```diff
@@ -0,0 +1,144 @@
+    register_group_permission_fields,
+    dispatch_uid="ephios.plugins.questionnaires.signals.register_group_permission_fields",
+)
+def group_permission_fields(sender, **kwargs):
+    return [
+        (
+            "manage_questions",
+            PermissionField(
+                label=_("Manage Questions"),
+                help_text=_("Enables this group to add questions that can be asked during signup."),
+                permissions=[
+                    "questionnaires.view_question",
+                    "questionnaires.add_question",
+                    "questionnaires.change_question",
+                    "questionnaires.delete_question",
+                ],
+            ),
+        )
+    ]
+
+
+@receiver(shift_forms, dispatch_uid="ephios.plugins.questionnaires.signals.shift_forms")
+def question_selection_form(sender, shift, request, **kwargs):
+    return [QuestionnaireForm(request.POST or None, shift=shift)]
+
+
+@receiver(signup_formfields, dispatch_uid="ephios.plugins.questionnaires.signals.signup_formfields")
+def provide_signup_formfields(
+    sender,
+    shift: Shift,
+    participant: AbstractParticipant,
+    participation: AbstractParticipation,
+    signup_choice,
+    **kwargs
+):
+    if signup_choice == BaseSignupForm.CHOICE_DECLINE:
+        return None
+
+    formfields = [
+        question.get_signup_formfield(participant, participation, signup_choice)
+        for question in Question.objects.filter(pk__in=shift.questionnaire)
+    ]
+
+    can_save_answers = isinstance(participant, LocalUserParticipant)
+    formfields.append(
+        AdditionalField(
+            "save_answers",
+            forms.BooleanField,
+            {
+                "label": _("Save answers to my user profile for future sign-ups"),
+                "initial": can_save_answers,
+                "disabled": not can_save_answers,
+                "required": False,
+                "widget": forms.HiddenInput if not can_save_answers else forms.CheckboxInput,
+            },
+            serializers.BooleanField,
+            {
+                "disabled": not can_save_answers,
+                "required": False,
+            },
+        )
+    )
+
+    return (
+        AdditionalFields(
+            "questionnaires",
+            {formfield.name: formfield for formfield in formfields},
+        )
+        if len(formfields) > 0
+        else None
+    )
+
+
+@receiver(signup_save, dispatch_uid="ephios.plugins.questionnaires.signals.signup_save")
+def save_signup(
+    sender,
+    shift: Shift,
+    participant: AbstractParticipant,
+    participation: AbstractParticipation,
+    cleaned_data,
+    **kwargs
+):
+    answers = {}
+
+    for name, value in cleaned_data.items():
+        if name == "questionnaires.save_answers" or not name.startswith("questionnaires."):
+            continue
+
+        # Field name format: questionnaires.123-question-name
+        question_pk = name.split(".", maxsplit=1)[1].split("-")[0]
```

Comment:
this seems a bit magical if one doesn't read the creation part next to it. maybe you can refactor this to a common spot/helper method?

Reply:
Thanks for the hint. Refactored in e69dceeb3f4e05817f9ab47d3b02433f7263d847